Allow linking a prebuilt optimized compiler-rt builtins library

pmur opened the pull request:
Extend the <target>.optimized-compiler-builtins bootstrap option to accept a path to a prebuilt compiler-rt builtins library, and update compiler-builtins to enable optimized builtins without building compiler-rt builtins.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

pmur committed:
Allow linking a prebuilt optimized compiler-rt builtins library

Extend the <target>.optimized-compiler-builtins bootstrap option to accept a
path to a prebuilt compiler-rt builtins library, and update compiler-builtins
to enable optimized builtins without building compiler-rt builtins.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16173867402/job/45654132079) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45654132079) [(plain)](https://github.com/rust-lang/rust/commit/27fcb13d14d5a3efccfc1ab35cc2af8a877bc73c/checks/45654132079/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Kobzol commented:
I don't really know much about the optimized compiler-rt builtins. Maybe @tgross35 might know more?

cuviper commented:
Right, it's less about time and more about wanting to use our existing system LLVM packages.

pmur commented:
Ping @tgross35. The intent here is to enable optimized-builtins in Fedora using the system's compiler-rt.builtins library.

tgross35 commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -567,25 +567,29 @@ pub fn std_cargo(builder: &Builder<'_>, target: TargetSelection, stage: u32, car
+        if let Some(path) = builder.config.optimized_compiler_builtins_path(target) {
+            cargo.env("LLVM_BUILTIN_RT_LIB", path);
+        } else {
+            // NOTE: this interacts strangely with `llvm-has-rust-patches`. In that case, we enforce `submodules = false`, so this is a no-op.
+            // But, the user could still decide to manually use an in-tree submodule.
+            //
+            // NOTE: if we're using system llvm, we'll end up building a version of `compiler-rt` that doesn't match the LLVM we're linking to.
+            // That's probably ok? At least, the difference wasn't enforced before. There's a comment in
+            // the compiler_builtins build script that makes me nervous, though:
+            // https://github.com/rust-lang/compiler-builtins/blob/31ee4544dbe47903ce771270d6e3bea8654e9e50/build.rs#L575-L579
```
Nit: rewrap comments to 100 chars

tgross35 commented on library/compiler-builtins/compiler-builtins/build.rs:
```diff
@@ -548,20 +548,28 @@ mod c {
+        // Optionally, link against a prebuilt compiler-rt library to supply
+        // optimized intrinsics instead of compiling a subset of compiler-rt
+        // from source.
+        let link_against_prebuilt_rt = env::var_os("LLVM_BUILTIN_RT_LIB").is_some();
```
Maybe `LLVM_COMPILER_RT_LIB` to match `RUST_COMPILER_RT_ROOT`, and since this is probably linking the entire compiler-rt rather than just builtins. Also I believe the LLVM library is called "builtins" rather than "builtin"

pmur replied:
In the case of fedora, compiler-rt builtins is packaged as a distinct library. I plan on linking against it. Looking at `build.rs`, on some targets (e.x openbsd), is only a broader compiler-rt library available to link against?

tgross35 reviewed (commented):
The changes to compiler-builtins lgtm with a possible rename of the var, assuming you have verified this works locally. Somebody more familiar with bootstrap (@Kobzol?) will need to review the changes there.

bors commented:
:umbrella: The latest upstream changes (presumably #144773) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pmur commented:
> > Is it a realistic case to consider something trying to link libgcc.a or libcompiler-rt.builtins.a when std already includes a copy of libcompiler-rt.builtins? That case I haven't tested.
> 
> I think this can happen in niche build setups, but no special handling should be needed; all three libraries make their builtin symbols weak, so conflicts aren't a problem.

The rust versions are weak, I am not sure the library built by optimized-compiler-builtins or the system provided libclang_rt.builtins.a exposes weak builtin functions. When I tried explicitly linking a second time on aarch64, it raised errors.

Kobzol reviewed (changes requested):
I took a look. Fine with the approach in general, but I'd appreciate two changes:
- It would be more consistent if also the `build.optimized-compiler-builtins` option took a path, so that we don't have the same named option that takes different values in `build` and in the `target` config.
- It would be nicer to create an enum, e.g. `CompilerBuiltinsConfig`, which would have two variants (`Build` and `UseExisting(PathBuf)`) and parse this enum from the TOML config files immediately, and then work with it in bootstrap, rather than converting the typed in `optimized_compiler_builtins`. That would also allow us to have only a single function in `Config`, and wouldn't require creating `optimized_compiler_builtins_path`.

pmur committed:
Refactor *.optimized-compiler-builtins bootstrap options

Create a dedicated enum to abstract the different ways compiler-builtins
can be configured.

This also relaxes build.optimized-compiler-builtins to accept the path
of a library to match the behavior of <target>.optimized-compiler-builtins
override.

Kobzol reviewed (commented):
Thank you, I think it's much easier to understand now! I'd appreciate some small comment changes, but other than that LGTM.

Kobzol reviewed (approved):
Since this changes config options, you could also add a change entry to `change_tracker.rs`, but this is probably so niche that it's not even needed.

LGTM from my side.

pmur commented:
Thanks. I also tried to improve the documentation in bootstrap.example.toml too.

bors commented:
:pushpin: Commit 148a07cbd0842ed64bffec0aad8b4c20ed1186a3 has been approved by `Kobzol,tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol,tgross35 148a07cbd0842ed64bffec0aad8b4c20ed1186a3 -->
<!-- homu: {"type":"Approved","sha":"148a07cbd0842ed64bffec0aad8b4c20ed1186a3","approver":"Kobzol,tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
